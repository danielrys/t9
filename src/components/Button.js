// @flow
import * as React from "react"
import { TouchableOpacity, Text, StyleSheet } from "react-native"

// theme
import { Colors, Metrics } from "../themes"

// types
import type { ViewType, TextType } from "../types"

const styles = StyleSheet.create({
  button: {
    alignItems: "center",
    borderRadius: Metrics.radii.base,
    height: Metrics.buttons.heights.base,
    justifyContent: "center",
  },
  primary: {
    backgroundColor: Colors.primary,
  },
  secondary: {
    backgroundColor: Colors.secondary,
  },
  text: {
    fontSize: Metrics.texts.sizes.base,
    color: Colors.background,
  },
})

type ButtonProps = {|
  onPress: () => void,
  children: string | React.Node,
  theme?: "primary" | "secondary",
  style?: ViewType,
  textStyle?: TextType,
|}

export default class Button extends React.PureComponent<ButtonProps> {
  static defaultProps = {
    style: {},
    textStyle: {},
    theme: "primary",
  }

  render() {
    const { children, onPress, theme, style, textStyle } = this.props
    return (
      <TouchableOpacity
        onPress={onPress}
        style={[styles.button, styles[theme || "primary"], style]}
      >
        {typeof children === "string" ? (
          <Text
            size={Metrics.texts.sizes.button}
            style={[styles.text, textStyle]}
          >
            {children}
          </Text>
        ) : (
          children
        )}
      </TouchableOpacity>
    )
  }
}
