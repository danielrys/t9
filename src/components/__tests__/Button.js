import React from "react"
import renderer from "react-test-renderer"

// components
import Button from "../Button"

describe("Button component", () => {
  it("renders with default props", () => {
    const tree = renderer
      .create(<Button>Default primary theme</Button>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
  it("renders with secondary theme", () => {
    const tree = renderer
      .create(
        <Button theme="secondary" onPress={() => null}>
          Secondary theme
        </Button>,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
  it("renders with disabled theme", () => {
    const tree = renderer
      .create(
        <Button theme="disabled" onPress={() => null}>
          Disabled theme
        </Button>,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
  it("renders with custom styles", () => {
    const tree = renderer
      .create(
        <Button
          style={{ backgroundColor: "red" }}
          textStyle={{ color: "yellow", fontSize: 12 }}
        >
          Button with custom styles
        </Button>,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
