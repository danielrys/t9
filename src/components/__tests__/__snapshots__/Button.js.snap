// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component renders with custom styles 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "red",
      "borderRadius": 10,
      "height": 84,
      "justifyContent": "center",
      "opacity": 1,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "rgb(255,255,255)",
          "fontSize": 16,
        },
        Object {
          "color": "yellow",
          "fontSize": 12,
        },
      ]
    }
  >
    Button with custom styles
  </Text>
</View>
`;

exports[`Button component renders with default props 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "rgb(24,100,167)",
      "borderRadius": 10,
      "height": 84,
      "justifyContent": "center",
      "opacity": 1,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "rgb(255,255,255)",
          "fontSize": 16,
        },
        Object {},
      ]
    }
  >
    Default primary theme
  </Text>
</View>
`;

exports[`Button component renders with disabled theme 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "borderRadius": 10,
      "height": 84,
      "justifyContent": "center",
      "opacity": 1,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "rgb(255,255,255)",
          "fontSize": 16,
        },
        Object {},
      ]
    }
  >
    Disabled theme
  </Text>
</View>
`;

exports[`Button component renders with secondary theme 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": undefined,
      "borderRadius": 10,
      "height": 84,
      "justifyContent": "center",
      "opacity": 1,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "rgb(255,255,255)",
          "fontSize": 16,
        },
        Object {},
      ]
    }
  >
    Secondary theme
  </Text>
</View>
`;
